### PR TITLE
Making 1.5.0 proof work end-to-end

### PR DESCRIPTION
**Pull request by mm-zk**
# What ❔

* All the necessary changes to make 1.5.0 proof work end to end (with new recursion tip and scheduler)
* This PR doesn't modify the verification keys (these are handled by separate PRs)

### CODE REVIEW

File: src/witness/individual_circuits/transient_storage_sorter.rs [hunk 1]
```diff
@@ -290,6 +290,10 @@ pub fn compute_transient_storage_dedup_and_sort<
         use circuit_definitions::encodings::log_query::transient_storage_comparison_key;
         let last_comparison_key = transient_storage_comparison_key(&last_sorted_query.raw_query);
         let last_timestamp = last_sorted_query.extended_timestamp;
+        let last_key = last_sorted_query.raw_query.key;
+        let last_address = last_sorted_query.raw_query.address;
+        let last_tx_number = last_sorted_query.raw_query.tx_number_in_block;
+        let last_shard_id = last_sorted_query.raw_query.shard_id;
```
Comment (EmilLuta) [inline src/witness/individual_circuits/transient_storage_sorter.rs:289]: This doesn't look right.

File: src/tests/complex_tests/mod.rs [hunk 2]
```diff
@@ -278,46 +282,77 @@ pub(crate) fn generate_base_layer(
...
-        //     }
-        //     _ => {
-        //         continue;
-        //     }
-        // }
-
-        base_test_circuit(el);
-    }
+    // HACK - added sorting.
+    recursion_queues.sort_by_key(|(circuit, _, _)| circuit.clone());
```
Comment (EmilLuta) [inline src/tests/complex_tests/mod.rs:334]: Most likely not needed.

File: src/tests/complex_tests/mod.rs [hunk 3]
```diff
@@ -278,46 +282,77 @@ pub(crate) fn generate_base_layer(
...
 
-    return;
+    if options.test_base_circuits {
```
Comment (EmilLuta) [inline src/tests/complex_tests/mod.rs:336]: I'd much rather move this in the other for loop.

File: src/tests/complex_tests/mod.rs [hunk 4]
```diff
@@ -331,24 +366,34 @@ fn run_and_try_create_witness_inner(
...
 
         if el.numeric_circuit_type() != previous_circuit_type {
             instance_idx = 0;
         }
 
-        if let Ok(proof) = source.get_base_layer_proof(el.numeric_circuit_type(), instance_idx) {
-            if instance_idx == 0 {
-                source.set_base_layer_padding_proof(proof).unwrap();
-            }
+        if options.reuse_artifacts {
```
Comment (EmilLuta) [inline src/tests/complex_tests/mod.rs]: You can't really reuse artifcats if not available.

```suggestion
        if options.try_reuse_artifacts {
```

File: src/tests/complex_tests/mod.rs [hunk 5]
```diff
@@ -965,10 +1035,18 @@ fn run_and_try_create_witness_inner(
...
         ..=(ZkSyncRecursionLayerStorageType::LeafLayerCircuitForEIP4844Repack as u8)
     {
-        let proof = source
-            .get_node_layer_proof(recursive_circuit_type, 0, 0)
-            .unwrap();
-        recursion_tip_proofs.push(proof.into_inner());
+        match source.get_node_layer_proof(recursive_circuit_type, 0, 0) {
+            Ok(proof) => recursion_tip_proofs.push(proof.into_inner()),
+            Err(_) => {
+                // HACK HACK HACK: reading the 'main VM' proof instead of the 'missing' one.
```
Comment (EmilLuta) [inline src/tests/complex_tests/mod.rs]: AFAIU, we actually can pass None's, based on @shamatar 's changes (I remember mentioning optional circuits).

Comment (mm-zk) [reply src/tests/complex_tests/mod.rs]: right now, this is still a 'vec proof ' and not 'vec optional proof '

File: src/tests/complex_tests/mod.rs (hunk 3 above)
Comment (mm-zk) [reply src/tests/complex_tests/mod.rs:336]: to which for loop?

File: .gitignore [hunk 6]
```diff
@@ -2,4 +2,9 @@
 Cargo.lock
 setup_2^26.key
 .idea
-.DS_Store
\ No newline at end of file
+.DS_Store
+
+# don't include the results of the testing (as they are 100M)
+test_proofs/base_layer
```
Comment (EmilLuta) [inline .gitignore:8]: Weren't there some of those already submitted? I believe yes, but if that's the case, let's delete them in this PR. If it was a different branch that I remember, just close this comment.

Comment (mm-zk) [reply .gitignore:8]: yes, they were - and now they are deleted (as they are too heavy).

File: src/compute_setups.rs [hunk 7]
```diff
@@ -588,6 +588,39 @@ mod test {
         generate_recursive_layer_vks_and_proofs(&mut source).expect("must compute setup");
     }
 
+    #[test]
```
Comment (EmilLuta) [inline src/compute_setups.rs:591]: Not a request for change, but I don't fully understand this test? Maybe it's my lack of knowledge on the codebase. Could you do me a TL;DR;? My biggest confusion here is that I don't see the circuit being created from scratch, but rather interacting with the `LocalFileDataSource` everywhere.

File: src/tests/complex_tests/mod.rs (hunk 3 above)
Comment (EmilLuta) [reply src/tests/complex_tests/mod.rs:336]: This [one](https://github.com/matter-labs/era-zkevm_test_harness/pull/125/files#diff-ccb4fd2bb9b63c6b5039bab735c14aac993009b69847360cfbd20c410f0e5856R376). They look the same to me, unless I'm missing anything obvious.

**Review by EmilLuta — APPROVED** [review]

**Pull request merged**